src/index.rs

```diff
@@ -340,61 +340,101 @@ impl ResourceIndex {
                     Err(_) => {
                         // a directory or empty file exists by the path
-                        self.forget_path(path, old_id)
+                        self.forget_path(canonical_path, old_id)
                     }
                     Ok(new_entry) => {
                         // valid resource exists by the path
-                        let curr_entry = &self.path2id[path];
+                        let curr_entry = &self.path2id[canonical_path];
 
                         if curr_entry.id == new_entry.id {
                             // in rare cases we are here due to hash collision
 
                             if curr_entry.modified == new_entry.modified {
-                                log::warn!("path {:?} was modified but not its content", &path);
+                                log::warn!("path {:?} was modified but not its content", &canonical_path);
                             }
 
-                            // the caller must have ensured that the path was updated
+                            // the caller must have ensured that the path was
+                            // updated
                             return Err(ArklibError::Collision(
                                 "New content has the same id".into(),
                             ));
                         }
 
                         // new resource exists by the path
-                        self.forget_path(path, old_id).map(|mut update| {
-                            update
-                                .added
-                                .insert(path_buf.clone(), new_entry.id);
-                            self.insert_entry(path_buf, new_entry);
-
-                            update
-                        })
+                        self.forget_path(canonical_path, old_id).map(
+                            |mut update| {
+                                update.added.insert(
+                                    canonical_path_buf.clone(),
+                                    new_entry.id,
+                                );
+                                self.insert_entry(
+                                    canonical_path_buf,
+                                    new_entry,
+                                );
+
+                                update
+                            },
+                        )
                     }
                 }
             }
         };
     }
 
+    pub fn forget_id(&mut self, old_id: ResourceId) -> Result<IndexUpdate> {
+        let old_path = self
+            .path2id
+            .drain()
+            .into_iter()
+            .filter_map(|(k, v)| {
+                if v.id == old_id {
+                    Some(k)
+                } else {
+                    None
+                }
+            })
+            .collect_vec();
+        for p in old_path {
+            self.path2id.remove(&p);
+        }
+        self.id2path.remove(&old_id);
```

Comment:
Probably we should handle collisions differently to remove necessity iterate through `path2id`. But this should be implemented as separate issue.

Reply:
- #58